### PR TITLE
Feature/conditional logic

### PR DESCRIPTION
**Pull request by timarney**
# Summary | Résumé

Demo 
https://www.youtube.com/watch?v=J0BhbZwNpso
 
-Updates the form builder to add a "custom rules" modal.
 The modal is available for elements that have pre determined values i.e. radios, checkboxes, and selects (dropdown).
 
 Example form file
 [example-form-2023-12-19.json](https://github.com/cds-snc/platform-forms-client/files/13717354/example-form-2023-12-19.json)

Screenshot 2023-12-28 at 12 27 54 PM

Screenshot 2023-12-14 at 8 57 52 AM

 - Adds a conditional wrapper around form elements to check for matching rules and groups
 - Adds a GC Forms provider to work with Matching rules and current values
 
 <hr>

## How it works
**Show / hide**

Screenshot 2023-12-21 at 11 40 09 AM

**Groups**

Screenshot 2023-12-21 at 11 40 09 AM

**Modal**

Screenshot 2023-12-21 at 11 43 42 AM

## Notes
- Logic for grouping exists and the schema has been updated but no UI is in place to create groups of elements.

### CODE REVIEW

File: public/static/locales/fr/form-builder.json [hunk 1]
```diff
@@ -699,5 +699,22 @@
...
+    "text1": "Show when: [fr]",
+    "text2": "Has value: [fr]",
+    "defaultQuestion": "No conditions always show [fr]",
+    "addCustomRules": "Add a rule [fr]",
+    "editCustomRules": "Edit rules [fr]",
+    "addAnotherRule": "Add another rule [fr]",
+    "removeRule": "remove [fr]",
+    "addOther": "Add 'if other' option [fr]",
+    "other": "Other [fr]",
+    "show": "Show: [fr]"
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:718]:
```suggestion
    "show": "Afficher :"
```

File: public/static/locales/fr/form-builder.json [hunk 2]
```diff
@@ -699,5 +699,22 @@
...
+    "questionTitle": "Show question: [fr]",
+    "text1": "Show when: [fr]",
+    "text2": "Has value: [fr]",
+    "defaultQuestion": "No conditions always show [fr]",
+    "addCustomRules": "Add a rule [fr]",
+    "editCustomRules": "Edit rules [fr]",
+    "addAnotherRule": "Add another rule [fr]",
+    "removeRule": "remove [fr]",
+    "addOther": "Add 'if other' option [fr]",
+    "other": "Other [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:717]:
```suggestion
    "other": "Autre ",
```

File: public/static/locales/en/form-builder.json [hunk 3]
```diff
@@ -699,5 +699,22 @@
...
+    "optionTitle": "If this option is selected:",
+    "questionTitle": "Show question:",
+    "text1": "Show when: ",
+    "text2": "Has value:",
+    "defaultQuestion": "No conditions always show",
+    "addCustomRules": "Add a rule",
+    "editCustomRules": "Edit rules",
+    "addAnotherRule": "Add another rule",
+    "removeRule": "remove",
+    "addOther": "Add 'if other' option",
```
Comment (anikbrazeau) [inline public/static/locales/en/form-builder.json:716]:
```suggestion
    "addOther": "Add input field for "other"",
```

File: public/static/locales/fr/form-builder.json [hunk 4]
```diff
@@ -699,5 +699,22 @@
...
+    "optionTitle": "If this option is selected: [fr]",
+    "questionTitle": "Show question: [fr]",
+    "text1": "Show when: [fr]",
+    "text2": "Has value: [fr]",
+    "defaultQuestion": "No conditions always show [fr]",
+    "addCustomRules": "Add a rule [fr]",
+    "editCustomRules": "Edit rules [fr]",
+    "addAnotherRule": "Add another rule [fr]",
+    "removeRule": "remove [fr]",
+    "addOther": "Add 'if other' option [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:716]:
```suggestion
    "addOther": "Ajouter un champ de saisi pour  « Autre »",
```

File: public/static/locales/fr/form-builder.json [hunk 5]
```diff
@@ -699,5 +699,22 @@
...
+    "modalTitle": "Add a rule to a question [fr]",
+    "modalTitleEdit": "Edit rules [fr]",
+    "modalDescription": "Allow an option in your question to show an additional short or long answer input field when selected. Only drop downs, single choice, and multi-select blocks can apply a rule. [fr]",
+    "optionTitle": "If this option is selected: [fr]",
+    "questionTitle": "Show question: [fr]",
+    "text1": "Show when: [fr]",
+    "text2": "Has value: [fr]",
+    "defaultQuestion": "No conditions always show [fr]",
+    "addCustomRules": "Add a rule [fr]",
+    "editCustomRules": "Edit rules [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:713]:
```suggestion
    "editCustomRules": "Modifier les règles",
```

File: public/static/locales/fr/form-builder.json [hunk 6]
```diff
@@ -699,5 +699,22 @@
...
+  "addConditionalRules": {
+    "modalTitle": "Add a rule to a question [fr]",
+    "modalTitleEdit": "Edit rules [fr]",
+    "modalDescription": "Allow an option in your question to show an additional short or long answer input field when selected. Only drop downs, single choice, and multi-select blocks can apply a rule. [fr]",
+    "optionTitle": "If this option is selected: [fr]",
+    "questionTitle": "Show question: [fr]",
+    "text1": "Show when: [fr]",
+    "text2": "Has value: [fr]",
+    "defaultQuestion": "No conditions always show [fr]",
+    "addCustomRules": "Add a rule [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:712]:
```suggestion
    "addCustomRules": "Ajouter une règle",
```

File: public/static/locales/fr/form-builder.json [hunk 7]
```diff
@@ -699,5 +699,22 @@
...
+    "editCustomRules": "Edit rules [fr]",
+    "addAnotherRule": "Add another rule [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:714]:
```suggestion
    "addAnotherRule": "Ajouter une autre règle",
```

File: public/static/locales/en/form-builder.json [hunk 8]
```diff
@@ -699,5 +699,22 @@
...
+    "other": "Other",
```
Comment (anikbrazeau) [inline public/static/locales/en/form-builder.json:717]:
```suggestion
    "other": "Other ",
```

File: public/static/locales/fr/form-builder.json [hunk 9]
```diff
@@ -699,5 +699,22 @@
...
+    "removeRule": "remove [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:715]:
```suggestion
    "removeRule": "Supprimer",
```

File: public/static/locales/en/form-builder.json [hunk 10]
```diff
@@ -699,5 +699,22 @@
...
+    "modalDescription": "Allow an option in your question to show an additional short or long answer input field when selected. Only drop downs, single choice, and multi-select blocks can apply a rule.",
+    "optionTitle": "If this option is selected:",
+    "questionTitle": "Show question:",
+    "text1": "Show when: ",
+    "text2": "Has value:",
+    "defaultQuestion": "No conditions always show",
+    "addCustomRules": "Add a rule",
+    "editCustomRules": "Edit rules",
+    "addAnotherRule": "Add another rule",
+    "removeRule": "remove",
```
Comment (anikbrazeau) [inline public/static/locales/en/form-builder.json:715]:
```suggestion
    "removeRule": "Remove",
```

File: public/static/locales/fr/form-builder.json [hunk 11]
```diff
@@ -699,5 +699,22 @@
     "description": "Veuillez télécharger une copie du formulaire et réessayer ou",
     "supportLink": "contacter l'équipe de soutien",
     "errorCode": "Code d'erreur : {{code}}"
+  },
+  "addConditionalRules": {
+    "modalTitle": "Add a rule to a question [fr]",
+    "modalTitleEdit": "Edit rules [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:705]:
```suggestion
    "modalTitleEdit": "Modifier les règles",
```

File: public/static/locales/fr/form-builder.json [hunk 12]
```diff
@@ -699,5 +699,22 @@
     "description": "Veuillez télécharger une copie du formulaire et réessayer ou",
     "supportLink": "contacter l'équipe de soutien",
     "errorCode": "Code d'erreur : {{code}}"
+  },
+  "addConditionalRules": {
+    "modalTitle": "Add a rule to a question [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:704]:
```suggestion
    "modalTitle": "Ajouter une règle à une question",
```

File: public/static/locales/fr/form-builder.json [hunk 13]
```diff
@@ -699,5 +699,22 @@
...
+    "modalTitleEdit": "Edit rules [fr]",
+    "modalDescription": "Allow an option in your question to show an additional short or long answer input field when selected. Only drop downs, single choice, and multi-select blocks can apply a rule. [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:706]:
```suggestion
    "modalDescription": "Afficher des champs de saisie supplémentaires en fonction de la sélection d'une option spécifique. Ceci s'applique aux questions à choix unique, à choix multiples ou à liste déroulante.",
```

File: public/static/locales/en/form-builder.json [hunk 14]
```diff
@@ -699,5 +699,22 @@
     "description": "Download a copy of the form and try again or",
     "supportLink": "contact Support",
     "errorCode": "Error code: {{code}}"
+  },
+  "addConditionalRules": {
+    "modalTitle": "Add a rule to a question",
+    "modalTitleEdit": "Edit rules",
+    "modalDescription": "Allow an option in your question to show an additional short or long answer input field when selected. Only drop downs, single choice, and multi-select blocks can apply a rule.",
```
Comment (anikbrazeau) [inline public/static/locales/en/form-builder.json:706]:
```suggestion
    "modalDescription": "Display additional input fields depending on the selection of a specific option. This applies to single choice, multiple choice, or dropdown list questions.",
```

File: public/static/locales/fr/form-builder.json [hunk 15]
```diff
@@ -699,5 +699,22 @@
...
+    "optionTitle": "If this option is selected: [fr]",
+    "questionTitle": "Show question: [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:708]:
```suggestion
    "questionTitle": "Afficher la question :",
```

File: public/static/locales/en/form-builder.json [hunk 16]
```diff
@@ -699,5 +699,22 @@
...
+    "optionTitle": "If this option is selected:",
```
Comment (anikbrazeau) [inline public/static/locales/en/form-builder.json:707]:
```suggestion
    "optionTitle": "If option selected is:",
```

File: public/static/locales/fr/form-builder.json [hunk 17]
```diff
@@ -699,5 +699,22 @@
     "description": "Veuillez télécharger une copie du formulaire et réessayer ou",
     "supportLink": "contacter l'équipe de soutien",
     "errorCode": "Code d'erreur : {{code}}"
+  },
+  "addConditionalRules": {
+    "modalTitle": "Add a rule to a question [fr]",
+    "modalTitleEdit": "Edit rules [fr]",
+    "modalDescription": "Allow an option in your question to show an additional short or long answer input field when selected. Only drop downs, single choice, and multi-select blocks can apply a rule. [fr]",
+    "optionTitle": "If this option is selected: [fr]",
```
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:707]:
```suggestion
    "optionTitle": "Si l'option sélectionnée est :",
```

File: public/static/locales/en/form-builder.json [hunk 18]
```diff
@@ -699,5 +699,22 @@
...
+    "questionTitle": "Show question:",
+    "text1": "Show when: ",
+    "text2": "Has value:",
+    "defaultQuestion": "No conditions always show",
```
Comment (anikbrazeau) [inline public/static/locales/en/form-builder.json:711]: @timarney Having a hard time finding where these strings show up while clicking around in the preview environment? Are they experimental for input field value conditionals?

File: public/static/locales/fr/form-builder.json (hunk 6 above)
Comment (anikbrazeau) [inline public/static/locales/fr/form-builder.json:712]:
```suggestion
    "text2": "Correspond à :",
    "defaultQuestion": "Aucune condition, sera toujours affiché",
    "addCustomRules": "Ajouter une règle",
```

File: public/static/locales/en/form-builder.json (hunk 18 above)
Comment (timarney) [reply public/static/locales/en/form-builder.json:711]: Might have been removed I'll double check in the A.M

**Review by thiessenp-cds — APPROVED** [review]
Lots of difficult logic solved in this PR. Nice work! LGTM :)

The UI interaction may need more explanation but user testing will show this if so

**Pull request merged**